File: java/collection.joookiwi/src/test/java/instance/GenericCollectionHolder_JoinToStringAlias.java
package instance;

import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Supplier;
import joookiwi.collection.java.GenericCollectionHolder;
import joookiwi.collection.java.callback.ObjIntFunction;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNullByDefault;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.Range;

import static joookiwi.collection.java.CommonContracts.ALWAYS_THIS_1;
import static joookiwi.collection.java.NumericConstants.MAX_INT_VALUE;
import static value.Arrays.ABCD;

@NotNullByDefault
public final class GenericCollectionHolder_JoinToStringAlias
        extends GenericCollectionHolder<String>
        implements StraightCollectionHolderForTest<String, GenericCollectionHolder_JoinToStringAlias> {

    /// The field that tell the amount of time the method `joinToString` has been called
    public @Range(from = 0, to = MAX_INT_VALUE) int amountOfCall = 0;

    public GenericCollectionHolder_JoinToStringAlias() { super(ABCD); }

    @Override public @Range(from = 0, to = MAX_INT_VALUE) int getAmountOfCall() { return amountOfCall; }

    @Contract(ALWAYS_THIS_1)
    @Override public GenericCollectionHolder_JoinToStringAlias execute(final Consumer<GenericCollectionHolder_JoinToStringAlias> action) {
        action.accept(this);
        return this;
    }


    @Override public String joinToString(final @Nullable String separator, final @Nullable String prefix, final @Nullable String postfix) {
        amountOfCall++;
        return super.joinToString(separator, prefix, postfix);
    }

    @Override public String joinToString(final @Nullable String separator, final @Nullable String prefix, final @Nullable String postfix, final int limit, final @Nullable String truncated) {
        amountOfCall++;
        return super.joinToString(separator, prefix, postfix, limit, truncated);
    }

    @Override public String joinToString(final @Nullable String separator, final @Nullable String prefix, final @Nullable String postfix, final int limit, final @Nullable String truncated, final @Nullable ObjIntFunction<? super String, String> transform) {
        amountOfCall++;
        return super.joinToString(separator, prefix, postfix, limit, truncated, transform);
    }

    @Override public String joinToString(final @Nullable String separator, final @Nullable String prefix, final @Nullable String postfix, final @Nullable Integer limit, final @Nullable String truncated, final @Nullable ObjIntFunction<? super String, String> transform) {
        amountOfCall++;
        return super.joinToString(separator, prefix, postfix, limit, truncated, transform);
    }

    @Override public String joinToString(final @Nullable String separator, final @Nullable String prefix, final @Nullable String postfix, final int limit, final @Nullable String truncated, final @Nullable Function<? super String, String> transform) {
        amountOfCall++;
        return super.joinToString(separator, prefix, postfix, limit, truncated, transform);
    }

    @Override public String joinToString(final @Nullable String separator, final @Nullable String prefix, final @Nullable String postfix, final @Nullable Integer limit, final @Nullable String truncated, final @Nullable Function<? super String, String> transform) {
        amountOfCall++;
        return super.joinToString(separator, prefix, postfix, limit, truncated, transform);
    }

    @Override public String joinToString(final @Nullable String separator, final @Nullable String prefix, final @Nullable String postfix, final int limit, final @Nullable String truncated, final @Nullable Supplier<String> transform) {
        amountOfCall++;
        return super.joinToString(separator, prefix, postfix, limit, truncated, transform);
    }

    @Override public String joinToString(final @Nullable String separator, final @Nullable String prefix, final @Nullable String postfix, final @Nullable Integer limit, final @Nullable String truncated, final @Nullable Supplier<String> transform) {
        amountOfCall++;
        return super.joinToString(separator, prefix, postfix, limit, truncated, transform);
    }

}

File: javascript/test/instance/GenericCollectionHolder_JoinToStringAlias.ts
//··························································
// Copyright (c) 2023-2026. Jonathan Bédard ~ JóôòKiwi
//
// This project is free to use.
// All the right is reserved to the author of this project.
// My projects:
//  - https://github.com/joooKiwi/type
//  - https://github.com/joooKiwi/lazy
//  - https://github.com/joooKiwi/collection
//  - https://github.com/joooKiwi/enumeration
//··························································

import type {Nullable, NullableNumber, NullableString} from "@joookiwi/type"

import type {StringCallback}                  from "../../src/type/callback"
import type {StraightCollectionHolderForTest} from "./StraightCollectionHolderForTest"

import {GenericCollectionHolder} from "../../src/GenericCollectionHolder"
import {ABCD}                    from "../value/arrays"

export class GenericCollectionHolder_JoinToStringAlias
    extends GenericCollectionHolder<string>
    implements StraightCollectionHolderForTest<string> {

    public amountOfCall = 0

    public constructor() { super(ABCD,) }

    public execute(action: (instance: this,) => void,): this {
        action(this,)
        return this
    }

    public override joinToString(separator?: NullableString, prefix?: NullableString, postfix?: NullableString, limit?: NullableNumber, truncated?: NullableString, transform?: Nullable<StringCallback<string>>,): string {
        this.amountOfCall++
        return super.joinToString(separator, prefix, postfix, limit, truncated, transform,)
    }

}
